File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>

using namespace std;

int N, cnt;
int map[15][15], col[15], diag[30], diag2[30];

void search(int y)
{
	if (y == N)
	{
		cnt++;
		return;
	}
	for (int i = 0; i < N; i++)
	{
		if (col[i] || diag[i + y] || diag2[i - y + N - 1])
		{
			continue;
		}
		col[i] = diag[i + y] = diag2[i - y + N - 1] = 1;
		search(y + 1);
		col[i] = diag[i + y] = diag2[i - y + N - 1] = 0;
	}
}

int main()
{
	cin >> N;

	search(0);
	cout << cnt;
	return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;
import java.util.*;

public class Main {        
    static int N;
    static boolean [][]map;
    static boolean [] row;
    static boolean [] col;
    static boolean [] diag;
    static boolean [] rediag;
    static int answer;
    static void nQueen(int cnt, int i) {
    	if(cnt == N) {
    		answer++;
    		return;
    	}
 
        for(int j=0; j <N ;j++) {
        	if(  !col[j] && !diag[i+j] && !rediag[N-1-(i-j)] ) {
        		col[j] = diag[i+j] = rediag[N-1-(i-j)] = true;
        		nQueen(cnt+1, i+1);
        		col[j] = diag[i+j] = rediag[N-1-(i-j)] = false;
        	}
        }
    }
    
    public static void main(String args[]) throws Exception
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());

        map = new boolean[N][N];
        
        row = new boolean[N];
        col = new boolean[N];
        diag = new boolean[2*N-1];
        rediag = new boolean[2*N-1];
              
        nQueen(0,0);
        System.out.println(answer);
    }
}
